### PR TITLE
fix(scripts): make install.ps1 work under PowerShell 7.x (pwsh)

### LINKED ISSUE #310: fix(scripts): make install.ps1 work under PowerShell 7.x (pwsh)
## Problem

`scripts/install.ps1` works under Windows PowerShell 5.1 (`powershell.exe`, the default on Windows Server 2019/2022) but fails when run under PowerShell 7.x (`pwsh.exe`) — for example on hosts that have removed the legacy 5.1 shell, or when an operator invokes the script explicitly via `pwsh -File install.ps1`. Tested against PowerShell 7.6.1; the same root cause applies to all 7.x versions.

The failure surfaces as a TLS-related download error when the script tries to resolve the latest alpamon release from `api.github.com`. Older Windows base images that default to TLS 1.0 / SSL 3 reach this state regularly; modern images may not reproduce it depending on OS-level TLS configuration.

### Root cause

The script forces TLS 1.2+ before any web request (`scripts/install.ps1:98-106`):

```powershell
# Force TLS 1.2+ before any Invoke-WebRequest / Invoke-RestMethod.
$tls = [Net.SecurityProtocolType]::Tls12
if ([Enum]::IsDefined([Net.SecurityProtocolType], 'Tls13')) {
    $tls = $tls -bor [Net.SecurityProtocolType]'Tls13'
}
[Net.ServicePointManager]::SecurityProtocol = $tls
```

This block is silently ineffective under PowerShell 7.x:

- **Windows PowerShell 5.1** (.NET Framework): `Invoke-WebRequest` and `Invoke-RestMethod` route through `ServicePointManager`. Assigning to `SecurityProtocol` directly affects TLS negotiation.
- **PowerShell 7.x** (.NET 7+/8+): `Invoke-WebRequest` / `Invoke-RestMethod` are re-implemented on top of `HttpClient`. `ServicePointManager.SecurityProtocol` still exists for source compatibility but **has no effect on actual TLS negotiation**. No warning, no error — the assignment silently no-ops.

The downstream consequence is that on a host where OS-level TLS defaults are TLS 1.0 / SSL 3, the subsequent `Invoke-RestMethod https://api.github.com/repos/alpacax/alpamon/releases/latest` performs the handshake with whatever the OS allows, and GitHub rejects it. The operator sees a generic TLS / HTTP error from `Invoke-RestMethod` even though the script "configured TLS 1.2+" successfully.

### Why this passed CI

`windows-latest` GitHub-hosted runners ship with PowerShell 5.1 enabled by default and have OS TLS already configured for 1.2+. The bug is invisible there.

## Expected behavior

- `install.ps1` succeeds end-to-end under both Windows PowerShell 5.1 and PowerShell 7.x on a host with modern OS TLS defaults.
- On a host with degraded OS TLS defaults (TLS 1.0 only), PowerShell 5.1 continues to force-upgrade negotiation to TLS 1.2+ as today. PowerShell 7.x does **not** attempt the no-op block; the script either succeeds (because OS TLS is fine and `HttpClient` selects TLS 1.2+) or fails with a clear, actionable error pointing at OS-level TLS configuration.

## Acceptance criteria

1. Wrap the TLS configuration block (`scripts/install.ps1:98-106`) in a PowerShell version guard:

   ```powershell
   # Force TLS 1.2+ before any Invoke-WebRequest / Invoke-RestMethod.
   # PowerShell 7.x uses HttpClient and ignores ServicePointManager;
   # rely on .NET / OS defaults there (TLS 1.2+ in .NET 7+).
   if ($PSVersionTable.PSVersion.Major -lt 7) {
       $tls = [Net.SecurityProtocolType]::Tls12
       if ([Enum]::IsDefined([Net.SecurityProtocolType], 'Tls13')) {
           $tls = $tls -bor [Net.SecurityProtocolType]'Tls13'
       }
       [Net.ServicePointManager]::SecurityProtocol = $tls
   }
   ```

2. Apply the same version guard to the `finally` restoration that puts `SecurityProtocol` back to `$previousSecurityProtocol` (`scripts/install.ps1` around line 229). Under PowerShell 7.x there is nothing to restore — skip both the save and the restore.

3. Add `#requires -Version 5.1` to the top of `install.ps1` so older PowerShell versions fail fast with a clear message rather than producing confusing errors mid-run. PowerShell 7.x passes this check trivially.

4. Add a short "PowerShell compatibility" subsection to `docs/windows.md`:

   > `install.ps1` works on Windows PowerShell 5.1 (`powershell.exe`) and PowerShell 7.x (`pwsh.exe`). On 5.1 the script forces TLS 1.2+ at the .NET level for compatibility with older OS images. On 7.x the script relies on the .NET 7+ HttpClient defaults (TLS 1.2 / 1.3); if the OS TLS configuration is older than that, configure it at the OS level before running the installer.

5. Smoke-test in CI under both shells. The existing `windows-latest` job already invokes the installer under 5.1; add a second invocation under `pwsh` (preinstalled on GitHub runners as 7.x) to exercise the new code path.

## Out of scope

- **Inspecting and configuring OS-level TLS** from the installer. That belongs to system administration, not to a per-host install script.
- **Switching `Invoke-RestMethod` to a manually constructed `HttpClient`** to force `SslProtocols` under PowerShell 7.x. The .NET 7+ default already negotiates TLS 1.2/1.3; the additional complexity is not justified for the failure modes we have actually observed.
- **Removing the 5.1 TLS block entirely**. Operators on older Server 2019 base images still depend on it.

## Dependencies

- ✅ alpamon Issue 1 (Windows CI) merged. The CI job already runs the installer once on `windows-latest`; the second invocation under `pwsh` is a small extension.

## Verification (manual)

1. **PowerShell 5.1 baseline**: on a Windows Server 2019 VM with default 5.1, run `.\install.ps1 -Url ... -Token ...`. Expect: success, no behavioral change from current main.
2. **PowerShell 7.x happy path**: on the same VM, install PowerShell 7.6.1 and run `pwsh -File .\install.ps1 -Url ... -Token ...`. Expect: success.
3. **PowerShell 7.x without legacy 5.1**: on a Windows host that has 5.1 disabled or removed, run via `pwsh`. Expect: success (no fallback to 5.1 attempted, no spurious TLS warning).
4. **Diagnostic check**: under PowerShell 7.x, confirm in the script output that the TLS block is not entered. A simple `Write-Verbose` line inside the guard helps debugging without affecting normal output.
5. **Regression**: confirm Linux and macOS code paths are unaffected. (The script is Windows-only, but the CI workflow should remain green for other OS matrix entries.)

## References (alpamon-side)

- `scripts/install.ps1:98-106` — the TLS `SecurityProtocol` block (the body of this fix)
- `scripts/install.ps1` `finally` block (around line 229) — the matching restoration to guard
- `.github/workflows/build-and-test.yml` / `.github/workflows/release.yml` — where to add the `pwsh` smoke run
- `docs/windows.md` — where to document PowerShell 5.1 vs 7.x compatibility

### PR DESCRIPTION
**Pull request by geunwoonoh**
## Summary

- `scripts/install.ps1` forces TLS 1.2+ via `[Net.ServicePointManager]::SecurityProtocol = $tls` so older Server 2019 images (TLS 1.0 / SSL 3 defaults) can complete the `api.github.com` handshake. Under PowerShell 7.x this silently no-ops because `Invoke-WebRequest` / `Invoke-RestMethod` are reimplemented on `HttpClient`, which ignores `ServicePointManager`. On hosts with degraded OS-level TLS, pwsh-launched installs fail with an opaque handshake error against `api.github.com`.
- Gate the entire save/set/restore triple on `$PSVersionTable.PSEdition -eq 'Desktop'` and rely on the HttpClient defaults under PS 7.x (TLS 1.2, plus TLS 1.3 when runtime and OS support it). Add `#requires -Version 5.1` so PS 3.x/4.x fail fast with a clear error. Add `Write-Verbose` markers on both branches for `-Verbose` diagnostics.
- Update `docs/windows.md` to differentiate the PS 5.1 (ServicePointManager shim) and PS 7.x (HttpClient defaults) paths and add a "PowerShell compatibility" subsection.
- Add two AST-parse-only CI steps under `shell: powershell` (5.1) and `shell: pwsh` (7.x) as a forward-looking regression guard for syntax compatibility.

Closes #310.

### `scripts/install.ps1` (4 edits, all sharing the guard `$PSVersionTable.PSEdition -eq 'Desktop'`)

1. **L1** — new `#requires -Version 5.1` directive at the top of the file.
2. **L82-101** — guard the **save**. Pre-initialize `$previousSecurityProtocol = $null` so the finally block has a defined read target under `Set-StrictMode -Version Latest`, then save the actual value only on PS 5.x.
3. **L114-138** — guard the **set**. Only run the TLS configuration block on PS 5.x; on PS 7.x emit a `Write-Verbose` saying we are deferring to `HttpClient` defaults.
4. **L255-269** — guard the **restore**. Symmetric with the save: only restore on PS 5.x so we never write `$null` (or anything else) into the `[Net.SecurityProtocolType]` property on PS 7.x.

Same guard expression is used in three places to keep `grep "PSEdition -eq 'Desktop'" scripts/install.ps1` returning exactly 3 hits — easy to audit for drift in future edits.

### `docs/windows.md` (2 edits)

1. **L85-93** — rewrite the now-stale prose `"TLS 1.2 is forced before any network call ..."` to differentiate PS 5.1 (ServicePointManager shim) from PS 7.x (HttpClient defaults), with a cross-link to the new section.
2. **L105-137** — new `### PowerShell compatibility` subsection (peer to `### Option A` / `### Option B`), with a 2-row matrix table, OS-TLS-degraded fallback guidance ("fix it at the OS level — installer is not the right place"), and a `-Verbose` diagnostic snippet.

### `.github/workflows/build-and-test.yml` (1 addition)

- Two new steps inside `build-and-test-windows`, between Build and Vet:
  - `shell: powershell` (5.1) → `Parser::ParseFile scripts\install.ps1`
  - `shell: pwsh` (7.x) → `Parser::ParseFile scripts/install.ps1`
- Failures emit `::error file=scripts/install.ps1,line=N::<message>` so the PR diff annotates the offending line directly.
- Since `release.yml` calls this workflow via `workflow_call`, the guards run on every release tag automatically too.

## Corrections to the source issue draft

While reviewing the draft (`GITHUB_ISSUE_ALPAMON_INSTALL_PS1_PWSH7.md`) against the actual code, five gaps were found and fixed in this PR:

1. **Issue missed guarding the SAVE at L85.** Without guarding it, `$previousSecurityProtocol` would be undefined under StrictMode or hold a meaningless value on PS 7.x. All three sites (save, set, restore) are now gated symmetrically.
2. **Issue claim that CI already invokes `install.ps1` under 5.1 is wrong.** Neither `build-and-test.yml` (only ran `--version` smoke on the built binary) nor `release.yml` (downloaded the release ZIP directly via `Invoke-WebRequest`) ever invoked `install.ps1`. The CI steps in this PR are net-new, not "a second invocation".
3. **Issue's docs change was purely additive but did not correct stale prose.** `docs/windows.md` L87-90 said `"TLS 1.2 is forced before any network call"` — true for 5.1, false for 7.x. The docs edit rewrites that sentence in addition to adding the new subsection.
4. **DEBUG doc mentioned `Write-Verbose` diagnostic lines; the issue draft omitted them.** Added — operators running `-Verbose` see which TLS path the installer took without any cost in normal runs.
5. **Issue's line range "L98-106" mixed 4 comment lines and 5 code lines.** Working precision matters; final commit references the executable code only.

## What this PR is honest about

- **CI proves syntax-compatibility regression only.** The AST parse step does not (and cannot, on GitHub-hosted `windows-latest` runners with healthy OS TLS) exercise the actual bug fix. Anyone claiming the new CI steps "prove the bug is fixed" is overselling.
- **The fix's correctness is established by manual V2 verification** on a host with degraded OS-level TLS — see Test plan below.

### Manual verification (required before merge)

- [ ] **V1 — PS 5.1 baseline (regression)**
  On Windows Server 2019 with default `powershell.exe` 5.1:
  ```powershell
  $env:ALPAMON_URL = "https://<workspace>"; $env:ALPAMON_TOKEN = "<valid-token>"
  powershell -ExecutionPolicy Bypass -File .\install.ps1 -Verbose
  ```
  Verbose output contains `Forced ServicePointManager.SecurityProtocol = ...` and `Restored ServicePointManager.SecurityProtocol = ...`. Exit 0; `Get-Service alpamon` reports `Running`.

- [ ] **V2 — PS 7.x on a degraded-TLS host (THE actual bug reproducer)**
  Same Server 2019 VM with `pwsh` 7.6.x and TLS 1.2 disabled at the OS level (older AMI).
  - Pre-fix sanity: on `git checkout main`, `pwsh -File .\install.ps1` should fail at the GitHub API handshake. If it succeeds, host is not in degraded state — pick a different AMI.
  - Fix branch: `pwsh -File .\install.ps1 -Verbose`. Expect: verbose output contains `Skipping ServicePointManager.SecurityProtocol assignment on PowerShell 7.x.x; HttpClient ignores it.` Exit 0; service `Running`.

- [ ] **V3 — `iwr | iex` cloud-init path on PS 7.x**
  Run the terser docs recipe at `docs/windows.md` `iwr ... | iex`. Expect success. The `$previousSecurityProtocol = $null` pre-init guards against state leak from the caller's scope.

- [ ] **V4 — `#requires` rejection**
  Realistic PS 4.0 test is not possible. Substitute: edit a scratch copy to `#requires -Version 99.0`, run, confirm friendly error message.

- [ ] **V5 — Docs render on GitHub**
  Cross-link `[PowerShell compatibility](#powershell-compatibility)` resolves, the new table renders, em-dashes have no spaces per project style.

### CI verification (automatic)

- [ ] **V6** — push triggers `Parse install.ps1 under Windows PowerShell 5.1` and `Parse install.ps1 under PowerShell 7.x` steps; both green. Linux container matrix and macOS jobs unaffected.

## Out of scope

Tracked elsewhere; not touched in this PR:

- Privilege demotion on Windows (alpamon `parsePaths` home guard / Issue 5 epic).
- `-DryRun` / `-DownloadOnly` flag for `install.ps1` to enable deeper CI smoke testing.
- End-to-end CI run against a real Alpacon endpoint (needs secrets + test workspace).
- Replacing `Invoke-RestMethod` with a manually constructed `HttpClient` to force `SslProtocols` under PS 7.x — modern HttpClient defaults already negotiate TLS 1.2 (and TLS 1.3 when available).
- Inspecting and configuring OS-level TLS from the installer.
- `Invoke-WebRequest -UseBasicParsing` deprecation under PS 7.x (currently a no-op, accepted without warning).
- Authenticode signing of `install.ps1`.

## Files changed

```
.github/workflows/build-and-test.yml | +43 / -0
docs/windows.md                      | +39 / -3
scripts/install.ps1                  | +44 / -6
3 files changed, 126 insertions(+), 9 deletions(-)
```

## References

- Issue #310 — source spec (this PR includes 5 corrections; see section above).
- Companion alpamon issue: Windows file download home-guard (DEBUG doc Issue 1) — separate ticket.
- `docs/windows.md` > [PowerShell compatibility](docs/windows.md#powershell-compatibility) — operator-facing docs.

### CODE REVIEW

File: scripts/install.ps1 [hunk 1]
```diff
@@ -82,7 +83,22 @@ $tempDir = $null
...
+# ServicePointManager.SecurityProtocol entirely, so there is nothing
+# to save or restore there. Pre-initialize to $null so the finally
+# block always has a defined read target even under
+# Set-StrictMode -Version Latest.
+$previousSecurityProtocol = $null
+if ($PSVersionTable.PSVersion.Major -lt 7) {
+    Write-Verbose "PowerShell $($PSVersionTable.PSVersion); saving previous SecurityProtocol for restore."
+    $previousSecurityProtocol = [Net.ServicePointManager]::SecurityProtocol
+}
+else {
```
[the remark this answers is not in the document]
Comment (geunwoonoh) [reply scripts/install.ps1]: Fixed. Switched the guard expression from `$PSVersionTable.PSVersion.Major -lt 7` to `$PSVersionTable.PSEdition -eq 'Desktop'` at all three sites (save, set, restore in commit 96b6746). PowerShell 6.x now correctly falls through to the HttpClient-defaults branch alongside 7.x. Surrounding comments were updated to refer to "Core editions (6.x and 7.x)" instead of just 7.x.

File: docs/windows.md [hunk 2]
```diff
@@ -99,6 +102,39 @@ $env:ALPAMON_TOKEN = "<TOKEN>"
...
+|---|---|---|
+| Windows PowerShell 5.1 | `powershell -ExecutionPolicy Bypass -File install.ps1` | `Invoke-WebRequest` routes through `[Net.ServicePointManager]`; the script forces TLS 1.2/1.3 there before any network call |
+| PowerShell 7.x | `pwsh -File install.ps1` | `Invoke-WebRequest` is reimplemented on `HttpClient`, which ignores `ServicePointManager.SecurityProtocol`; the script relies on the .NET 7+ defaults (TLS 1.2/1.3) |
+
+The version check uses `$PSVersionTable.PSVersion.Major -lt 7`, so the
+5.1 code path also runs under the (end-of-life) PowerShell 6.x line if
+it is somehow still in use. PowerShell 4.x and earlier are rejected by
+a `#requires -Version 5.1` directive at the top of the script, which
+fails fast with a clear error instead of producing a confusing mid-run
+cmdlet failure.
```
[the remark this answers is not in the document]
Comment (geunwoonoh) [reply docs/windows.md]: Fixed. Rewrote the paragraph below the matrix to describe the new `$PSVersionTable.PSEdition -eq 'Desktop'` check and clarify that PS 6.x and 7.x both report `Core` edition and take the HttpClient-defaults path. The matrix table still names the two supported targets (5.1 and 7.x); the prose now covers the 6.x edge case explicitly.

File: .github/workflows/build-and-test.yml [hunk 3]
```diff
@@ -153,6 +153,49 @@ jobs:
         run: go build -v .
         working-directory: ./cmd/alpamon
 
+      # Parse install.ps1 under both Windows PowerShell 5.1 and
+      # PowerShell 7.x to catch future syntax-compatibility drift.
+      # This is a parse-only regression guard, not a behavior test:
+      # the install.ps1 PS-7.x fix gates ServicePointManager assignments
+      # on $PSVersionTable.PSVersion.Major, which both shells parse
```
[the remark this answers is not in the document]
Comment (geunwoonoh) [reply .github/workflows/build-and-test.yml]: Fixed. Updated the comment block in `.github/workflows/build-and-test.yml` (commit 5659188) so it now references `$PSVersionTable.PSEdition (only the Desktop edition takes the legacy shim)` — matches the iteration 1 guard change.

File: scripts/install.ps1 [hunk 4]
```diff
@@ -82,7 +83,23 @@ $tempDir = $null
...
+# there is nothing to save or restore there. Pre-initialize to
+# $null so the finally block always has a defined read target
+# even under Set-StrictMode -Version Latest.
+$previousSecurityProtocol = $null
+if ($PSVersionTable.PSEdition -eq 'Desktop') {
+    Write-Verbose "PowerShell $($PSVersionTable.PSVersion); saving previous SecurityProtocol for restore."
+    $previousSecurityProtocol = [Net.ServicePointManager]::SecurityProtocol
+}
+else {
+    Write-Verbose "PowerShell $($PSVersionTable.PSVersion); skipping SecurityProtocol save (HttpClient ignores it)."
```
[the remark this answers is not in the document]
Comment (geunwoonoh) [reply scripts/install.ps1:101]: Fixed. The PR description was edited via `gh pr edit` to refer to `$PSVersionTable.PSEdition -eq 'Desktop'` in all three spots (the save/set/restore guard mention, the heading, and the `grep` audit hint). The code and description now agree.

File: docs/windows.md [hunk 5]
```diff
@@ -99,6 +102,43 @@ $env:ALPAMON_TOKEN = "<TOKEN>"
...
+
+`install.ps1` is verified against both Windows PowerShell 5.1
+(`powershell.exe`, the in-box shell on Windows Server 2019/2022 and
+Windows 10/11) and PowerShell 7.x (`pwsh.exe`, the cross-platform
+build shipped via the Microsoft Store and `winget`).
+
+| Shell | Invocation | TLS path |
+|---|---|---|
+| Windows PowerShell 5.1 | `powershell -ExecutionPolicy Bypass -File install.ps1` | `Invoke-WebRequest` routes through `[Net.ServicePointManager]`; the script forces TLS 1.2/1.3 there before any network call |
+| PowerShell 7.x | `pwsh -File install.ps1` | `Invoke-WebRequest` is reimplemented on `HttpClient`, which ignores `ServicePointManager.SecurityProtocol`; the script relies on the .NET 7+ defaults (TLS 1.2/1.3) |
```
[the remark this answers is not in the document]
Comment (geunwoonoh) [reply docs/windows.md]: Fixed. Rephrased the PowerShell 7.x table cell to "HttpClient/OS defaults (TLS 1.2, plus TLS 1.3 when the runtime and SChannel/OpenSSL support it — on Windows that means Server 2022 / Windows 11 or newer)" so an operator on Server 2019 with pwsh 7.x is not misled into expecting TLS 1.3.

File: scripts/install.ps1 [hunk 6]
```diff
@@ -99,11 +116,27 @@ try {
...
+    #
+    # PowerShell Core editions (6.x and 7.x) reimplement
+    # Invoke-WebRequest / Invoke-RestMethod on top of HttpClient;
+    # ServicePointManager.SecurityProtocol still exists for source
+    # compatibility but has no effect on TLS negotiation there.
+    # Skip the assignment on those editions and rely on the
+    # HttpClient defaults (TLS 1.2 / 1.3 on .NET 7+). If OS-level
+    # TLS is degraded below 1.2 on such a host, fix it at the OS
+    # level; the installer is not the right place to paper over
+    # that.
```
[the remark this answers is not in the document]
Comment (geunwoonoh) [reply scripts/install.ps1]: Fixed (commit 04f508d). Reworded the inline comment to "HttpClient/OS defaults (TLS 1.2, plus TLS 1.3 when the runtime and OS support it — on Windows that requires Server 2022 / Windows 11 SChannel or newer; older Windows gets TLS 1.2 only)". Dropped the ".NET 7+" reference since PS 7.0/7.1/7.2 ship on earlier .NET runtimes.

File: docs/windows.md [hunk 7]
```diff
@@ -99,6 +102,43 @@ $env:ALPAMON_TOKEN = "<TOKEN>"
...
+### PowerShell compatibility
+
+`install.ps1` is verified against both Windows PowerShell 5.1
+(`powershell.exe`, the in-box shell on Windows Server 2019/2022 and
+Windows 10/11) and PowerShell 7.x (`pwsh.exe`, the cross-platform
+build shipped via the Microsoft Store and `winget`).
+
+| Shell | Invocation | TLS path |
+|---|---|---|
+| Windows PowerShell 5.1 | `powershell -ExecutionPolicy Bypass -File install.ps1` | `Invoke-WebRequest` routes through `[Net.ServicePointManager]`; the script forces TLS 1.2/1.3 there before any network call |
```
[the remark this answers is not in the document]
Comment (geunwoonoh) [reply docs/windows.md]: Fixed (commit 4faa0ee). Reworded the Windows PowerShell 5.1 cell to "forces TLS 1.2 (and TLS 1.3 when the `Tls13` enum is present, which requires .NET Framework 4.8+) before any network call". This matches what the script actually does — the `[Enum]::IsDefined` probe at install.ps1:127-129 means stock Server 2019 images on .NET 4.7.x end up on TLS 1.2 only. Symmetric with the iteration 3 fix to the 7.x row.

**Review by eunyoung14 — APPROVED** [review]
